### DISCUSSION TITLE
Electron Preload - Context Bridge event to Quasar

### DISCUSSION THREAD

**Discussion by SonoranBrian** [quasarframework/quasar#10033] [CLI - Electron mode]
I've upgraded my project to v2, but am having a hard time with the electron context bridge. V1 allowed me to listen directly to the ipcRenderer right in a Vue file.

Specifically, I'm looking to have the electron main app inform a Vue file when a hotkey is pressed.

```js
// electron-preload.js
import { contextBridge, ipcRenderer } from 'electron'

const validChannels = ['hotKeyPressed'];
contextBridge.exposeInMainWorld(
    'myWindowAPI', {
        registerKeys: (keys) => {
            // Sends hotkey config to main process for registration
            ipcRenderer.send('registerKeys', keys);
        },
    on: (channel, func) => {
        // Debug - This never gets hit?
        console.log('got keypress from main');
        if (validChannels.includes(channel)) {
            // Send event for Vue
            // How to emit an event back up?
        }
      },
    },
);
```

```js
// electron-main.js
ipcMain.on('registerKeys', (e, config) => {
... some processing ...
    // Register keys
    keys.forEach(key => {
      globalShortcut.register(`${prefix}+${key}`, () => {
        // Works properly
        console.log(`${key} is pressed`)
        // Send keypress back to preload file for the context bridge?
        mainWindow.webContents.send('hotKeyPressed', key);
      })
    });
  });
```

```js
// somefile.vue
// Send hotkey config to electron - Works
window.myWindowAPI.registerKeys(/*some data*/);

// Handle hotkey press in Vue file
window.myWindowAPI.on('hotKeyPressed', (key) => {
  console.log('Handle hotkey press in Vue');
});
```

I can send data from the Vue UI just fine via
`window.myWindowAPI.registerKeys(/*some data*/);`
That reaches the contextBridge in the preload just fine, and registers the keys in the main electron file - but I can't seem to get any sort of event from the main electron process (`electron-main.js`) to get back to the context bridge (`electron-preload.js`) so I can listen to it in a Vue file (`somefile.vue`).

1. How can the electron-main.js emit an event back up to the contextBridge in electron-preload.js? Is `mainWindow.webContents.send('hotKeyPressed', key);` correct?

2. How can the contextBridge in `electron-preload.js` listen to that event from `electron-main.js` and send it up to a Vue file?

I've been trying several different methods with no luck on getting the preload file to listen for the event from electron-main - any examples or help here would be amazing.

**Comment by SonoranBrian** [comment]
Alrighty, hopefully this can help someone else out.
A working example of events to/from Vue to the electron-preload and electron-main. This is specifically used in my app to register global hotkeys in electron and handle them in Vue.

```js
// somefile.vue
// SEND event to electron-preload.js contextBridge
      window.myWindowAPI.registerKeys(/*some data*/);

// RECIEVE events from electron-main.js through the electron-preload.js contextBridge
      window.myWindowAPI.on('hotKeyPressed', (key) => {
        // Handle hotkey press in Vue file
        this.handleHotKey(key);
      });
```
```js
// electron-preload.js
import { contextBridge, ipcRenderer } from 'electron'

const validChannels = ['hotKeyPressed'];
contextBridge.exposeInMainWorld(
    'myWindowAPI', {
        registerKeys: (keys) => {
            // Sends hotkey config to main process for registration
            ipcRenderer.send('registerKeys', keys);
        },
    on: (channel, func) => {
        if (validChannels.includes(channel)) {
            // Deliberately strip event as it includes `sender`
            ipcRenderer.on(channel, (event, ...args) => func(...args))
        }
      },
    },
);
```
```js
// electron-main.js
app.on('ready', function()  {
/* ... */

  // electron main handling for registerKeys event send from Vue through the preload's context bridge
  ipcMain.on('registerKeys', (e, config) => {

    // Unregister all existing shortcuts (Update)
    globalShortcut.unregisterAll()

    keys.forEach(key => {
      globalShortcut.register(`${prefix}+${key}`, () => {
        console.log(`${key} is pressed`)
        // Send hotKeyPressed event up to preload contextBridge
        mainWindow.webContents.send('hotKeyPressed', key);
      })
    });
  });
});
```

**Reply by stanles75** [reply]
I was just trying to figure out how to do this and came across your post.  I'm going to give this a shot in my own project.  Definitely helpful!

**Comment by hawkeye64** [comment]
@SonoranBrian @stanles75 
You can also compare my Electron project: https://github.com/hawkeye64/electron-quasar-file-explorer-v2

**Comment by kevinwt95** [comment]
Thanks! This should definitely be included in the official documentation as an example
